rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,881 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 136 - Modern Signature Algorithms
+
+## Required Approvers
+
+* Engineering: (@jakule || @espadolini)
+* Security: (@reedloden || @jentfoo) && Doyensec
+* Product: @klizhentas
+
+## What
+
+Teleport should support modern key types and signature algorithms, currently
+only RSA2048 keys are supported with the PKCS#1 v1.5 signature scheme.
+This applies to CA keys and subject (user/host/db) keys, but each can/will be
+addressed individually.
+
+## Why
+
+Modern algorithms like ECDSA and Ed25519 offer better security properties with
+smaller keys that are faster to generate and sign with.
+Some of the more restrictive security policies are starting to reject RSA2048
+(e.g. [RHEL 8's FUTURE policy](https://access.redhat.com/articles/3642912)).
+
+## Details
+
+### Summary
+
+We will introduce a new config to `teleport.yaml` and `cluster_auth_preference`
+to control the key types and signature algorithms used by Teleport CAs and all
+clients and hosts which have certificates issued by those CAs.
+
+This config will default to a `recommended` set of algorithms for each protocol
+chosen by us to balance security, compatibility, and performance.
+We will reserve the right to change this set of `recommended` algorithms when
+either:
+
+* the major version of the auth server's teleport.yaml config changes, or
+* in a major version release of Teleport.
+
+Most Teleport administrators will never need to see or interact with this config
+because they can trust that we will select a vetted set of standards-compliant
+algorithms that are trusted to be secure, and we will not break compatibility
+with internal Teleport components or third-party software unless deemed
+absolutely necessary for security reasons.
+
+Teleport administrators will be able to deviate from the `recommended`
+algorithms when they have a compliance need (they must use a particular
+algorithm) or a compatibility need (one of our selected algorithms is not
+supported by an external softare that interacts with Teleport in their
+deployment).
+
+Here is what the config will look like in its default state:
+
+```yaml
+ca_key_params:
+  user:
+    ssh:
+      algorithm: recommended
+      allowed_subject_algorithms: [recommended]
```

Comment:
mentioned in the HSM section

>  We will attempt to use the configured CA algorithms when the CA keys are backed
by HSMs or KMS services.
If the specific algorithm is not supported, we will do our best to return an
informative error message to the user.
It will be the reponsibility of the Teleport admin to select an algorithm
supported by their particular HSM/KMS.